src/Language/PureScript/CoreFn/Convert/DesugarObjects.hs

```diff
@@ -0,0 +1,533 @@
+
+       allVars :: Expr Ann -> [(SourceType,Qualified Ident)]
+       allVars ex = ex ^.. icosmos @Context @(Expr Ann) M.empty . _Var . to (\(_,b,c) -> (b,c))
+
+       findBoundVar :: Ident -> Expr Ann -> Maybe (SourceType, Qualified Ident)
+       findBoundVar nm ex = find (goFind . snd) (allVars ex)
+         where
+           goFind = \case
+             Qualified (ByModuleName _) _ -> False
+             Qualified ByNullSourcePos _ -> False -- idk about this actually, guess we'll find out
+             Qualified (BySourcePos _) nm' -> nm == nm'
+
+       goList :: (Expr Ann -> Expr Ann) -> [Expr Ann] -> Either ExprConvertError [Exp FVar]
+       goList _ [] = pure []
+       goList g (ex:exs) = do
+         e' <- go g ex
+         es' <- goList g exs
+         pure $ e' : es'
+
+       -- ONLY USE THIS IN LAMBDA ABSTRACTIONS!
+       matchVar :: Ty -> Ident -> FVar -> Maybe BVar
+       matchVar t nm (FVar ty n')
+         | ty == t && nm == n' =  Just (BVar 0 t nm)
+         | otherwise = Nothing
+
+       tryConvertLit :: (Expr Ann -> Expr Ann) -> Literal (Expr Ann) -> Either ExprConvertError (Either (Exp FVar) (Lit (Exp FVar)))
+       tryConvertLit cb = \case
+         NumericLiteral (Left i) -> pure . Right $ IntL i
+         NumericLiteral (Right d) -> pure . Right $ NumL d
+         StringLiteral pss -> pure . Right $ StringL pss
+         CharLiteral c     -> pure . Right $ CharL c
+         BooleanLiteral b  -> pure . Right $ BoolL b
+         ArrayLiteral xs   -> Right . ArrayL <$> traverse (go cb) xs
+         ObjectLiteral fs'  -> do -- TODO Maybe check for empty? I think that's a legal expr?
+           let fs = sortOn fst fs'
+               len = length fs
+               fakeCName = mkFakeCName len
+               fakeTName = mkFakeTName len
+               bareFields = snd <$> fs
+               types' = exprType <$> bareFields
+               types = types' <> [foldl' srcTypeApp (srcTypeConstructor fakeTName) types']
+               ctorType = foldr1 function types
+               ctorExp = Constructor nullAnn ctorType (disqualify fakeTName) (disqualify fakeCName) []
+           ctor <-  assembleDesugaredObjectLit ctorExp ctorType bareFields
+           Left <$> go cb ctor
+
+       goType :: SourceType -> Either ExprConvertError Ty
+       goType = catchTE . tryConvertType
+
+       catchTE :: forall t. Either TypeConvertError t -> Either ExprConvertError t
+       catchTE = first ((\x -> ExprConvertError f expression x "Failed to convert type") . Just)
+
+assembleDesugaredObjectLit :: Expr Ann -> SourceType -> [Expr Ann] -> Either ExprConvertError (Expr Ann)
+assembleDesugaredObjectLit expr (_ :-> b) (arg:args) = assembleDesugaredObjectLit (App nullAnn b expr arg) b args
+assembleDesugaredObjectLit expr _ [] = pure expr -- TODO better error
+assembleDesugaredObjectLit _ _ _ = error "something went wrong in assembleDesugaredObjectLit"
+
+desugarObjectAccessor :: SourceType -> PSString -> Expr Ann -> Maybe (Expr Ann)
```

Comment:
Could you please add a small illustration in a comment above every function like this? Perhaps containing some semi-real PS code snippet